### PR TITLE
ROX-19955: Prometheus resources in sensor upgrader

### PR DESCRIPTION
**Pull request by stehessel**
## Description

The sensor upgrader is used when sensor is installed via manifest download, while at a later point central is updated to a newer version. The upgrader will then attempt to upgrade sensor to the same upgraded version. As part of the upgrade process, it currently checks that all resources are created in the stackrox namespace. With `4.2.0`, when integration with OpenShift monitoring is enabled, sensor additionally creates Prometheus resources in system namespaces.

This leads to a failure during the upgrader pre-flight check when upgrading from `4.1.x` to `4.2.0` for OpenShift secured clusters. As a fix for this issue, this PR adds the Prometheus resources to an exception list.

## Testing Performed

To reproduce the original issue:
* Install central via Helm chart version `4.1.3`.
* Install sensor via manifest download by running `./sensor.sh`.
* Upgrade central to `4.2.0` via `helm upgrade`.
* In the cluster overview page, the sensor upgrader shows an error in the pre-flight check.

To test the fix:
* Install central via Helm chart version `4.1.3`.
* Install sensor via manifest download by running `./sensor.sh`.
* Upgrade central to `4.2.0-fix` via `helm upgrade`.
* In the cluster overview page, the sensor upgrader successfully performs the upgrade.

### CODE REVIEW

File: sensor/upgrader/preflight/namespace.go [hunk 1]
```diff
@@ -1,20 +1,40 @@
...
+	"openshift-monitoring": {"ServiceMonitor", "PrometheusRule"},
+}
+
 type namespaceCheck struct{}
 
 func (namespaceCheck) Name() string {
 	return "Allowed namespaces"
 }
 
+func isException(resource *k8sobjects.ObjectRef) bool {
```
Comment (dhaus67) [inline sensor/upgrader/preflight/namespace.go:22]:
```suggestion
func matchesException(resource *k8sobjects.ObjectRef) bool {
```

File: sensor/upgrader/preflight/namespace.go [hunk 2]
```diff
@@ -1,20 +1,40 @@
...
+		}
+	}
+	return false
+}
+
 func (namespaceCheck) Check(_ *upgradectx.UpgradeContext, execPlan *plan.ExecutionPlan, reporter checkReporter) error {
 	for _, act := range execPlan.Actions() {
-		if act.ObjectRef.Namespace != "" && act.ObjectRef.Namespace != common.Namespace {
+		act := act
+		doCheck := !isException(&act.ObjectRef)
```
Comment (dhaus67) [inline sensor/upgrader/preflight/namespace.go:36]:
```suggestion
		noException := !isException(&act.ObjectRef)
```

**Review by dhaus67 — CHANGES_REQUESTED** [review]
We should have a unit test covering the exception conditions.

Also, is there someone else that definitely _needs_ to take a look at that, like Maple / Draco?

**Review by fredrb — COMMENTED** [review]
The current change seems ok. But I wonder why are we implementing a fine-grained exception list rather than changing the premise that things can only be installed in `common.Namespace`? 

I don't have the historical context because this was here long before I joined, but my understanding is that we want to make sure that things weren't installed in the incorrect namepsace. I.e.: if the upgrader had incorrectly determined resources to be installed elsewhere, but stackrox should be confined to `stackrox` namespace. 

Should't we change this premise, and say that ACS can be installed across a set of namespaces, like `stackrox`, `kube-system` and `openshift-monitoring`, without a distinction for which resource type is allowed in each? 

The reasoning for this is to avoid exactly the bug this PR is fixing in the future -- where we need a new resource to be installed, and need to patch sensor to support that.

**Comment by fredrb** [comment]
By the way, disregard `ci/prow/ocp-4-10-sensor-profiling`.

**Comment by stehessel** [comment]
> The current change seems ok. But I wonder why are we implementing a fine-grained exception list rather than changing the premise that things can only be installed in common.Namespace?

Are you suggesting to simply drop the namespace pre-flight check? I would agree that it doesn't server much purpose, because the resources are defined by the Helm chart anyway, which the upgrader simply renders. The upgrader doesn't even decide the resource definitions / namespaces, but is testing for them.

The reason I didn't drop the check is that I wanted the most minimal fix, but indeed maybe it's better to drop completely.

**Comment by stehessel** [comment]
Decided to keep the namespace check for the scope of this fix to reduce the chance of further regression. It's not 100% clear that we can simply drop the namespace check altogether, although I do believe that it is likely the case. However, that would require some more thorough investigation, which I don't think should block the fix.

@fredrb If you like I can create a ticket to properly remove the namespace check and file it under Sensor tech debt.

**Review by dhaus67 — APPROVED** [review]

**Comment by stehessel** [comment]
@dhaus67 Sorry to re-ping, but could you take another look? I actually found another location in which the upgrader explicitly assumes that resources live in certain namespaces. I extended this list to the OpenShift monitoring case and commented on how to extend this to other resources in the future - if it ever comes to that.

**Review by fredrb — APPROVED** [review]

**Review by dhaus67 — APPROVED** [review]

**Pull request merged**